src/gui/plugins/apply_force_torque/ApplyForceTorque.cc

```diff
@@ -0,0 +1,336 @@
+ *
+*/
+
+#include <mutex>
+#include <vector>
+#include <string>
+
+#include <gz/gui/Application.hh>
+#include <gz/gui/MainWindow.hh>
+#include <gz/sim/gui/GuiEvents.hh>
+#include <gz/sim/Link.hh>
+#include <gz/sim/Model.hh>
+#include <gz/sim/World.hh>
+#include <gz/sim/EntityComponentManager.hh>
+#include <gz/sim/components/World.hh>
+#include <gz/sim/components/Name.hh>
+#include <gz/msgs/Utility.hh>
+#include <gz/msgs/entity_wrench.pb.h>
+#include <gz/plugin/Register.hh>
+#include <gz/transport/Node.hh>
+
+#include "ApplyForceTorque.hh"
+
+namespace gz
+{
+namespace sim
+{
+  class ApplyForceTorquePrivate
+  {
+    /// \brief Publish wrench messages
+    public: void PublishWrench(bool _applyForce, bool _applyTorque);
+
+    /// \brief Transport node
+    public: transport::Node node;
+
+    /// \brief Publisher for EntityWrench messages
+    public: transport::Node::Publisher pub;
+
+    /// \brief To synchronize member access
+    public: std::mutex mutex;
+
+    /// \brief Name of the selected model
+    public: QString modelName;
+
+    /// \brief List of the name of the links in the selected model
+    public: QStringList linkNameList;
+
+    /// \brief List of links in the selected model
+    public: std::vector<Link> linkList;
+
+    /// \brief Index of selected link in list
+    public: int linkIndex{-1};
+
+    /// \brief Currently selected entities, organized by order of selection.
+    public: std::vector<Entity> selectedEntities;
+
+    /// \brief True if a new entity was selected
+    public: bool changedEntity{false};
+
+    /// \brief Force to be applied
+    public: math::Vector3d force{0.0, 0.0, 0.0};
+
+    /// \brief Offset of the force application relative to center of mass
+    public: math::Vector3d offset{0.0, 0.0, 0.0};
+
+    /// \brief Torque to be applied
+    public: math::Vector3d torque{0.0, 0.0, 0.0};
+  };
+}
+}
+
+using namespace gz;
+using namespace sim;
+
+/////////////////////////////////////////////////
+ApplyForceTorque::ApplyForceTorque()
+  : GuiSystem(), dataPtr(std::make_unique<ApplyForceTorquePrivate>())
+{
+}
+
+/////////////////////////////////////////////////
+ApplyForceTorque::~ApplyForceTorque() = default;
+
+/////////////////////////////////////////////////
+void ApplyForceTorque::LoadConfig(const tinyxml2::XMLElement */*_pluginElem*/)
+{
+  if (this->title.empty())
+    this->title = "Apply force and torque";
+
+  gz::gui::App()->findChild<gz::gui::MainWindow *>
+      ()->installEventFilter(this);
+  gz::gui::App()->findChild<gz::gui::MainWindow *>
+      ()->QuickWindow()->installEventFilter(this);
+}
+
+/////////////////////////////////////////////////
+bool ApplyForceTorque::eventFilter(QObject *_obj, QEvent *_event)
+{
+  if (_event->type() ==
+    gz::sim::gui::events::EntitiesSelected::kType)
+  {
+    gz::sim::gui::events::EntitiesSelected *_e =
+        static_cast<gz::sim::gui::events::EntitiesSelected*>(_event);
+    this->dataPtr->selectedEntities = _e->Data();
```

Comment:
If we only need one selected entity, can we just store that instead of the vector? If you need to know whether the selected entity has been set or not, you can use `std::optional`.